plugins/uefi-capsule/fu-uefi-grub-device.c

```diff
@@ -0,0 +1,199 @@
+	if (grub_reboot == NULL) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_FAILED,
+				    "could not find grub-reboot");
+		return FALSE;
+	}
+
+	/* replace ESP info in conf with what we detected */
+	g_string_append_printf(str, "EFI_PATH=%s\n", target_app);
+	fu_common_string_replace(str, esp_path, "");
+	g_string_append_printf(str, "ESP=%s\n", esp_path);
+	grub_target = g_build_filename(localstatedir, "uefi_capsule.conf", NULL);
+	if (!g_file_set_contents(grub_target, str->str, -1, error))
+		return FALSE;
+
+	/* refresh GRUB configuration */
+	argv_mkconfig[0] = grub_mkconfig;
+	if (!g_spawn_sync(NULL,
+			  (gchar **)argv_mkconfig,
+			  NULL,
+			  G_SPAWN_DEFAULT,
+			  NULL,
+			  NULL,
+			  &output,
+			  NULL,
+			  NULL,
+			  error))
+		return FALSE;
+	if (g_getenv("FWUPD_UPDATE_CAPSULE_VERBOSE") != NULL)
+		g_debug("%s", output);
+
+	/* make fwupd default */
+	argv_reboot[0] = grub_reboot;
+	return g_spawn_sync(NULL,
+			    (gchar **)argv_reboot,
+			    NULL,
+			    G_SPAWN_DEFAULT,
+			    NULL,
+			    NULL,
+			    NULL,
+			    NULL,
+			    NULL,
+			    error);
+}
+
+static gboolean
+fu_uefi_grub_device_write_firmware(FuDevice *device,
+				   FuFirmware *firmware,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	FuUefiDevice *self = FU_UEFI_DEVICE(device);
+	const gchar *fw_class = fu_uefi_device_get_guid(self);
+	g_autofree gchar *basename = NULL;
+	g_autofree gchar *directory = NULL;
+	g_autofree gchar *esp_path = fu_uefi_device_get_esp_path(self);
+	g_autofree gchar *fn = NULL;
+	g_autofree gchar *source_app = NULL;
+	g_autofree gchar *target_app = NULL;
+	g_autofree gchar *varname = fu_uefi_device_build_varname(self);
+	g_autoptr(GBytes) fixed_fw = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	/* ensure we have the existing state */
+	if (fw_class == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "cannot update device info with no GUID");
+		return FALSE;
+	}
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* save the blob to the ESP */
+	directory = fu_uefi_get_esp_path_for_os(device, esp_path);
+	basename = g_strdup_printf("fwupd-%s.cap", fw_class);
+	fn = g_build_filename(directory, "fw", basename, NULL);
+	if (!fu_common_mkdir_parent(fn, error))
+		return FALSE;
+	fixed_fw = fu_uefi_device_fixup_firmware(self, fw, error);
+	if (fixed_fw == NULL)
+		return FALSE;
+	if (!fu_common_set_contents_bytes(fn, fixed_fw, error))
+		return FALSE;
+
+	/* delete the logs to save space; use fwupdate to debug the EFI binary */
+	if (fu_efivar_exists(FU_EFIVAR_GUID_FWUPDATE, "FWUPDATE_VERBOSE")) {
+		if (!fu_efivar_delete(FU_EFIVAR_GUID_FWUPDATE, "FWUPDATE_VERBOSE", error))
+			return FALSE;
+	}
+	if (fu_efivar_exists(FU_EFIVAR_GUID_FWUPDATE, "FWUPDATE_DEBUG_LOG")) {
+		if (!fu_efivar_delete(FU_EFIVAR_GUID_FWUPDATE, "FWUPDATE_DEBUG_LOG", error))
+			return FALSE;
+	}
```

Comment:
I think you want these to only run in non self tests (IE move them later in the function).

Reply:
Or move the tests for FWUPD_UEFI_TEST up? Do you think we want to run fu_uefi_device_write_update_info() and fu_uefi_device_fixup_firmware() for example?

Reply:
Fixed, thanks.